### PR TITLE
[OIDC 6] Find matching federated credential policy (policy evaluation)

### PR DESCRIPTION
**Pull request by joelverhagen**
Progress on https://github.com/NuGet/NuGetGallery/issues/10212.
Depends on https://github.com/NuGet/NuGetGallery/pull/10268.

This is a big chunk of the OIDC feature's business logic. A new `IFederatedCredentialEvaluator` is added which takes a list of federated credential policies and a bearer token and evaluates which trust policy, if any matches the given bearer token.

This is a read-only operation. The creation of the short-lived API key will be performed at a higher level.

There are two main parts of the logic.

First, find the issuer of the bearer token and validate it using the appropriate, issuer-specific validation routine. This step can be summarized as "Is this a valid token, issued by an expected issuer?". This is done no matter the policies provided. to reduce the amount of information leaked about policies in our DB.

Second, if the token is valid, the code checks the claims in the token against the given polices. For the short term, this is checking if the `tid` and `oid` claims in a policy match the token.

For a future phase where we accept GitHub Actions tokens, the first step will be enhanced with GitHub OIDC signature validation and step 2 will validate against some expected build context provided by the user (e.g. is the token coming from an expected repository and branch). This is not part of the PR but I wanted to give a taste for how this code will expand in the future as we federate with more IdPs.

The returned `EvaluatedFederatedCredentialPolicies` is the result pattern because there are several error modes that should be handled differently, but gracefully. This is opposed to throwing different flavored exceptions, Generally, there is "bad token", "no matching policy", and "matching policy". The "bad token" state shows a helpful error message because it is unrelated to any specific policy. This eases debugging for the user. The "unauthorized" state will not show a detailed error message because we don't want to leak information about what policies the user has or doesn't have.

The caller of this new method will fetch the policies for a given user and call this method to find a matching policy if any.

### CODE REVIEW

File: src/NuGetGallery.Services/Authentication/Federated/EntraIdServicePrincipalCriteria.cs [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
+#nullable enable
+
+namespace NuGetGallery.Services.Authentication
+{
+    public class EntraIdServicePrincipalCriteria
+    {
+        [JsonConstructor]
+        public EntraIdServicePrincipalCriteria(Guid tenantId, Guid objectId)
+        {
+            TenantId = tenantId == Guid.Empty ? throw new ArgumentNullException(nameof(tenantId)) : tenantId;
```
Comment (agr) [inline src/NuGetGallery.Services/Authentication/Federated/EntraIdServicePrincipalCriteria.cs:16]: `ArgumentOutOfRangeException`?

Comment (joelverhagen) [inline src/NuGetGallery.Services/Authentication/Federated/EntraIdServicePrincipalCriteria.cs:16]:
```suggestion
            TenantId = tenantId == Guid.Empty ? throw new ArgumentOutOfRangeException(nameof(tenantId)) : tenantId;
```

File: src/NuGetGallery.Services/Authentication/Federated/EntraIdServicePrincipalCriteria.cs [hunk 2]
```diff
@@ -0,0 +1,26 @@
...
+
+namespace NuGetGallery.Services.Authentication
+{
+    public class EntraIdServicePrincipalCriteria
+    {
+        [JsonConstructor]
+        public EntraIdServicePrincipalCriteria(Guid tenantId, Guid objectId)
+        {
+            TenantId = tenantId == Guid.Empty ? throw new ArgumentOutOfRangeException(nameof(tenantId)) : tenantId;
+            ObjectId = objectId == Guid.Empty ? throw new ArgumentNullException(nameof(objectId)) : objectId;
```
Comment (joelverhagen) [inline src/NuGetGallery.Services/Authentication/Federated/EntraIdServicePrincipalCriteria.cs:17]:
```suggestion
            ObjectId = objectId == Guid.Empty ? throw new ArgumentOutOfRangeException(nameof(objectId)) : objectId;
```

File: src/NuGetGallery.Services/Authentication/Federated/EntraIdServicePrincipalCriteria.cs (hunk 1 above)
Comment (joelverhagen) [reply src/NuGetGallery.Services/Authentication/Federated/EntraIdServicePrincipalCriteria.cs]: Done.

File: src/NuGetGallery.Services/Authentication/Federated/FederatedCredentialEvaluator.cs [hunk 3]
```diff
@@ -0,0 +1,282 @@
...
+            };
+
+            _logger.LogInformation(validationException, "The JSON web token with recognized issuer {Issuer} could not be validated.", issuer);
+
+            return (userError, null);
+        }
+        
+        private async Task<(string? UserError, string? Identifier, TokenValidationResult? Result)> GetEntraIdValidationResultAsync(JsonWebToken jwt)
+        {
+            if (!jwt.TryGetPayloadValue<string>("uti", out var uti)
```
Comment (agr) [inline src/NuGetGallery.Services/Authentication/Federated/FederatedCredentialEvaluator.cs:217]: Consider creating human readable constants for JWT properties.

Comment (joelverhagen) [reply src/NuGetGallery.Services/Authentication/Federated/FederatedCredentialEvaluator.cs]: Done.

**Comment by zhhyu** [comment]
Looking...

File: src/NuGetGallery.Services/Authentication/Federated/EvaluatedFederatedCredentialPolicies.cs [hunk 4]
```diff
@@ -0,0 +1,90 @@
...
+            => new(EvaluatedFederatedCredentialPoliciesType.NoMatchingPolicy, userError: null, results);
+    }
+
+    public enum FederatedCredentialPolicyResultType
+    {
+        Success, // policy matched token
+        Unauthorized, // details are hidden from the user
+    }
+
+    public class FederatedCredentialPolicyResult
```
Comment (zhhyu) [inline src/NuGetGallery.Services/Authentication/Federated/EvaluatedFederatedCredentialPolicies.cs:62]: nit: maybe it helps if this public class is moved to a separate FederatedCredentialPolicyResult.cs. The class is also used in the evaluator, not just EvaluatedFederatedCredentialPolicies. Not sure though.

File: src/NuGetGallery.Services/Authentication/Federated/EvaluatedFederatedCredentialPolicies.cs [hunk 5]
```diff
@@ -0,0 +1,90 @@
...
+    {
+        private readonly FederatedCredential? _federatedCredential;
+        private readonly string? _internalReason;
+
+        public FederatedCredentialPolicyResult(
```
Comment (zhhyu) [inline src/NuGetGallery.Services/Authentication/Federated/EvaluatedFederatedCredentialPolicies.cs:67]: >public [](http://example.com/codeflow?start=8&length=6)

Is this a private constructor?

File: src/NuGetGallery.Services/Authentication/Federated/FederatedCredentialEvaluator.cs [hunk 6]
```diff
@@ -0,0 +1,311 @@
...
+            // sort the policy results by creation date, so older policy results are preferred
+            foreach (var policy in policies.OrderBy(x => x.Created))
+            {
+                var result = EvaluatePolicy(policy, validatedJwt);
+
+                results.Add(result);
+
+                if (result.Type == FederatedCredentialPolicyResultType.Success)
+                {
+                    return EvaluatedFederatedCredentialPolicies.MatchedPolicy(results, result.Policy, result.FederatedCredential);
```
Comment (zhhyu) [inline src/NuGetGallery.Services/Authentication/Federated/FederatedCredentialEvaluator.cs:66]: >results [](http://example.com/codeflow?start=78&length=7)

If 1 policy is matched, are other already evaluated policies used anywhere?

File: src/NuGetGallery.Services/Authentication/Federated/EvaluatedFederatedCredentialPolicies.cs [hunk 7]
```diff
@@ -0,0 +1,90 @@
...
+            EvaluatedFederatedCredentialPoliciesType type,
+            string? userError = null,
+            IReadOnlyList<FederatedCredentialPolicyResult>? results = null,
+            FederatedCredentialPolicy? policy = null,
+            FederatedCredential? federatedCredential = null)
+        {
+            Type = type;
+            _userError = userError;
+            _results = results;
+            _policy = policy;
```
[the remark this answers is not in the document]
Comment (joelverhagen) [reply src/NuGetGallery.Services/Authentication/Federated/EvaluatedFederatedCredentialPolicies.cs:36]: Good idea, I will do that.

File: src/NuGetGallery.Services/Authentication/Federated/EvaluatedFederatedCredentialPolicies.cs (hunk 4 above)
Comment (joelverhagen) [reply src/NuGetGallery.Services/Authentication/Federated/EvaluatedFederatedCredentialPolicies.cs:62]: Sure thing!

File: src/NuGetGallery.Services/Authentication/Federated/EvaluatedFederatedCredentialPolicies.cs (hunk 5 above)
Comment (joelverhagen) [reply src/NuGetGallery.Services/Authentication/Federated/EvaluatedFederatedCredentialPolicies.cs:67]: This should be private, yes. I am using the result pattern and expect only static factory methods to be used. Good catch.

File: src/NuGetGallery.Services/Authentication/Federated/FederatedCredentialEvaluator.cs (hunk 6 above)
Comment (joelverhagen) [reply src/NuGetGallery.Services/Authentication/Federated/FederatedCredentialEvaluator.cs:66]: The next policies after the first matched will be skipped. We prefer the first created so the behavior is deterministic. The data model is pretty generic (N policies can be owned by a user) so we need to handle the duplicate case somehow.

Comment (zhhyu) [reply src/NuGetGallery.Services/Authentication/Federated/FederatedCredentialEvaluator.cs:66]: That makes sense. I am not sure how the caller uses results. If there is 1 matched policy, it seems that other evaluated but not matched polices are not really needed for the caller.

Comment (joelverhagen) [reply src/NuGetGallery.Services/Authentication/Federated/FederatedCredentialEvaluator.cs:66]: Yeah, given the order of my PRs it is unclear what the caller may do. In short the caller (a new `FederatedCredentialService`) will fetch the policies for a given user. If any policy matches, a temporary API key will be created, linked to the policy. The policy will have it's "last used" date updated and a DB foreign key will be added to the create API key. This allows deletion of the policy to clear related API keys.

Comment (joelverhagen) [reply src/NuGetGallery.Services/Authentication/Federated/FederatedCredentialEvaluator.cs:66]: This will be clear in a future PR.

**Review by zhhyu — APPROVED** [review]

**Pull request merged**